state/indexer/block/kv/util.go

```diff
@@ -51,6 +51,51 @@ func heightKey(height int64) ([]byte, error) {
 	)
 }
 
+func keyBelongsToHeightRange(key []byte, left, right int64) bool {
+	// left included, right excluded
+	eventHeight, err := parseHeightFromEventKey(key)
+	if err == nil {
+		return left <= eventHeight && eventHeight < right
+	}
+
+	var (
+		blockHeightKeyPrefix string
+		possibleHeight       int64
+	)
+	remaining, err := orderedcode.Parse(string(key), &blockHeightKeyPrefix, &possibleHeight)
+	if err != nil {
+		return false
+	}
+	if len(remaining) == 0 &&
+		blockHeightKeyPrefix == types.BlockHeightKey &&
+		left <= possibleHeight &&
+		possibleHeight < right {
+		return true
+	}
+	return false
+}
+
+func getHeightFromKey(key []byte) int64 {
+	// Must be called with either heightKey or eventKey
+	eventHeight, err := parseHeightFromEventKey(key)
+	if err == nil {
+		return eventHeight
+	}
+
+	var (
+		blockHeightKeyPrefix string
+		possibleHeight       int64
+	)
+	remaining, err := orderedcode.Parse(string(key), &blockHeightKeyPrefix, &possibleHeight)
+	if err != nil {
+		panic(err)
+	}
+	if len(remaining) == 0 && blockHeightKeyPrefix == types.BlockHeightKey {
+		return possibleHeight
+	}
+	panic(fmt.Errorf("key must be either heightKey or eventKey"))
```

Comment:
Are you not storing also the retain heights in the stores? This means that the iterator will hit them and it will cause a panic?

Reply:
This function is meant to be called only with either heightKey or eventKey. I tried to emphasize it with [this](https://github.com/cometbft/cometbft/blob/7810d2654880cfcaf532fae234b82fc041b67e3c/state/indexer/block/kv/util.go#L79) annotation

This is ensured [here](https://github.com/cometbft/cometbft/blob/7810d2654880cfcaf532fae234b82fc041b67e3c/state/indexer/block/kv/kv.go#L146)

Reply:
I am not sure I see how this call enforces that it is called only for a key containing height? the iterator should get all keys unless you are somehow checking that the key has a height?

Reply:
Clarified in person.